src/ott/solvers/nn/neuraldual.py

```diff
@@ -108,55 +146,135 @@ def setup(
-      neural_g.pos_weights = self.pos_weights
+    warn_str = f"Setting of ICNN and the positive weights setting of the " \
+        f"`W2NeuralDual` are not consistent. Proceeding with " \
+        f"the `W2NeuralDual` setting, with positive weights " \
+        f"being {self.pos_weights}."
+    if isinstance(neural_f, icnn.ICNN):
+      if neural_f.pos_weights != self.pos_weights:
+        warnings.warn(warn_str)
+        neural_f.pos_weights = self.pos_weights
+
+    if isinstance(neural_g, icnn.ICNN):
+      if neural_g.pos_weights != self.pos_weights:
+        warnings.warn(warn_str)
+        neural_g.pos_weights = self.pos_weights
 
     self.state_f = neural_f.create_train_state(rng_f, optimizer_f, input_dim)
     self.state_g = neural_g.create_train_state(rng_g, optimizer_g, input_dim)
 
-    # define train and valid step functions
-    self.train_step_f = self.get_step_fn(train=True, to_optimize="f")
-    self.valid_step_f = self.get_step_fn(train=False, to_optimize="f")
+    # Assume g defines the potential unless this attribute is set.
+    self.g_returns_potential = neural_g.returns_potential \
+        if hasattr(neural_g, 'returns_potential') else False
 
-    self.train_step_g = self.get_step_fn(train=True, to_optimize="g")
-    self.valid_step_g = self.get_step_fn(train=False, to_optimize="g")
+    if self.num_inner_iters == 1 and self.parallel_updates:
+      self.train_step_parallel = self.get_step_fn(
+          train=True, to_optimize="both"
+      )
+      self.valid_step_parallel = self.get_step_fn(
+          train=False, to_optimize="both"
+      )
+      self.train_fn = self.train_neuraldual_parallel
+    else:
+      if self.parallel_updates:
+        warnings.warn(
+            'parallel_updates set to True but disabling it because num_inner_iters>1'
+        )
+      self.train_step_f = self.get_step_fn(train=True, to_optimize="f")
+      self.valid_step_f = self.get_step_fn(train=False, to_optimize="f")
+      self.train_step_g = self.get_step_fn(train=True, to_optimize="g")
+      self.valid_step_g = self.get_step_fn(train=False, to_optimize="g")
+      self.train_fn = self.train_neuraldual_alternating
 
   def __call__(
       self,
       trainloader_source: Iterable[jnp.ndarray],
       trainloader_target: Iterable[jnp.ndarray],
       validloader_source: Iterable[jnp.ndarray],
       validloader_target: Iterable[jnp.ndarray],
-  ) -> Union[potentials.DualPotentials, Tuple[potentials.DualPotentials,
-                                              Train_t]]:
-    logs = self.train_neuraldual(
+      callback: Optional[Callable] = None,
+  ) -> Union[Potentials_t, Tuple[Potentials_t, Train_t]]:
+    logs = self.train_fn(
         trainloader_source,
         trainloader_target,
         validloader_source,
         validloader_target,
+        callback=callback,
     )
     res = self.to_dual_potentials()
 
     return (res, logs) if self.logging else res
 
-  def train_neuraldual(
+  def train_neuraldual_parallel(
```

Comment:
I originally decided to keep these separate since the conditionals necessary to merge them would decrease how easy it is to read/edit them, but maybe it's not so bad. I'll keep this unresolved for now and add it in a larger editing pass soon